File: webapp/go/user_handler.go
package main

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"net/http"
	"os/exec"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/sessions"
	"github.com/jmoiron/sqlx"
	"github.com/labstack/echo-contrib/session"
	"github.com/labstack/echo/v4"
	"golang.org/x/crypto/bcrypt"
)

const (
	defaultSessionIDKey      = "SESSIONID"
	defaultSessionExpiresKey = "EXPIRES"
	defaultUserIDKey         = "USERID"
	defaultUserNameKey       = "USERNAME"
	bcryptDefaultCost        = bcrypt.MinCost
)

type UserModel struct {
	ID          int64  `db:"id"`
	Name        string `db:"name"`
	DisplayName string `db:"display_name"`
	Description string `db:"description"`
	// HashedPassword is hashed password.
	HashedPassword string `db:"password"`
	// CreatedAt is the created timestamp that forms an UNIX time.
	CreatedAt int64 `db:"created_at"`
	UpdatedAt int64 `db:"updated_at"`
}

type User struct {
	ID          int64  `json:"id"`
	Name        string `json:"name"`
	DisplayName string `json:"display_name,omitempty"`
	Description string `json:"description,omitempty"`
	// CreatedAt is the created timestamp that forms an UNIX time.
	CreatedAt int64 `json:"created_at,omitempty"`
	UpdatedAt int64 `json:"updated_at,omitempty"`

	Theme Theme `json:"theme,omitempty"`
}

type Theme struct {
	ID        int64 `json:"id"`
	DarkMode  bool  `json:"dark_mode"`
	CreatedAt int64 `json:"created_at"`
}

type ThemeModel struct {
	ID        int64 `db:"id"`
	UserID    int64 `db:"user_id"`
	DarkMode  bool  `db:"dark_mode"`
	CreatedAt int64 `db:"created_at"`
}

type PostUserRequest struct {
	Name        string `json:"name"`
	DisplayName string `json:"display_name"`
	Description string `json:"description"`
	// Password is non-hashed password.
	Password string               `json:"password"`
	Theme    PostUserRequestTheme `json:"theme"`
}

type PostUserRequestTheme struct {
	DarkMode bool `json:"dark_mode"`
}

type LoginRequest struct {
	UserName string `json:"username"`
	// Password is non-hashed password.
	Password string `json:"password"`
}

func getMeHandler(c echo.Context) error {
	ctx := c.Request().Context()

	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		return err
	}

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}
	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	userModel := UserModel{}
	err = tx.GetContext(ctx, &userModel, "SELECT * FROM users WHERE id = ?", userID)
	if errors.Is(err, sql.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, err.Error())
	}
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	user, err := fillUserResponse(ctx, tx, userModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, user)
}

// ユーザ登録API
// POST /user
func registerHandler(c echo.Context) error {
	ctx := c.Request().Context()
	defer c.Request().Body.Close()

	req := PostUserRequest{}
	if err := json.NewDecoder(c.Request().Body).Decode(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "failed to decode the request body as json")
	}

	if req.Name == "pipe" {
		return echo.NewHTTPError(http.StatusBadRequest, "the username 'pipe' is reserved")
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcryptDefaultCost)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "begin tx failed")
	}
	defer tx.Rollback()

	now := time.Now().Unix()
	userModel := UserModel{
		Name:           req.Name,
		DisplayName:    req.DisplayName,
		Description:    req.Description,
		HashedPassword: string(hashedPassword),
		CreatedAt:      now,
		UpdatedAt:      now,
	}

	result, err := tx.NamedExecContext(ctx, "INSERT INTO users (name, display_name, description, password, created_at, updated_at) VALUES(:name, :display_name, :description, :password, :created_at, :updated_at)", userModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "user insertion failed")
	}

	userID, err := result.LastInsertId()
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "last insert id failed")
	}

	userModel.ID = userID

	themeModel := ThemeModel{
		UserID:    userID,
		DarkMode:  req.Theme.DarkMode,
		CreatedAt: now,
	}
	if _, err := tx.NamedExecContext(ctx, "INSERT INTO themes (user_id, dark_mode, created_at) VALUES(:user_id, :dark_mode, :created_at)", themeModel); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "theme insertion failed")
	}

	if out, err := exec.Command("pdnsutil", "add-record", "u.isucon.dev", req.Name, "A", "30", powerDNSSubdomainAddress).CombinedOutput(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, string(out))
	}

	user, err := fillUserResponse(ctx, tx, userModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to fill user response")
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "commit failed")
	}

	return c.JSON(http.StatusCreated, user)
}

// ユーザログインAPI
// POST /login
func loginHandler(c echo.Context) error {
	ctx := c.Request().Context()
	defer c.Request().Body.Close()

	req := LoginRequest{}
	if err := json.NewDecoder(c.Request().Body).Decode(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "failed to decode the request body as json")
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	userModel := UserModel{}
	// usernameはUNIQUEなので、whereで一意に特定できる
	err = tx.GetContext(ctx, &userModel, "SELECT * FROM users WHERE name = ?", req.UserName)
	if errors.Is(err, sql.ErrNoRows) {
		return echo.NewHTTPError(http.StatusUnauthorized, "invalid username or password")
	}
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	err = bcrypt.CompareHashAndPassword([]byte(userModel.HashedPassword), []byte(req.Password))
	if err == bcrypt.ErrMismatchedHashAndPassword {
		return echo.NewHTTPError(http.StatusUnauthorized, "invalid username or password")
	}
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	sessionEndAt := time.Now().Add(1 * time.Hour)

	sessionID := uuid.NewString()

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	sess.Options = &sessions.Options{
		Domain: "u.isucon.dev",
		MaxAge: int(60000 /* 10 seconds */), // FIXME: 600
		Path:   "/",
	}
	sess.Values[defaultSessionIDKey] = sessionID
	// FIXME: ユーザ名
	sess.Values[defaultUserIDKey] = userModel.ID
	sess.Values[defaultUserNameKey] = userModel.Name
	sess.Values[defaultSessionExpiresKey] = sessionEndAt.Unix()

	if err := sess.Save(c.Request(), c.Response()); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusOK)
}

// ユーザ詳細API
// GET /user/:username
func getUserHandler(c echo.Context) error {
	ctx := c.Request().Context()
	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		return err
	}

	username := c.Param("username")

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	userModel := UserModel{}
	if err := tx.GetContext(ctx, &userModel, "SELECT * FROM users WHERE name = ?", username); err != nil {
		return echo.NewHTTPError(http.StatusNotFound, err.Error())
	}

	user, err := fillUserResponse(ctx, tx, userModel)
	if err != nil {
		return err
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, user)
}

func getUsersHandler(c echo.Context) (err error) {
	ctx := c.Request().Context()

	if err := verifyUserSession(c); err != nil {
		return err
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	var userModels []*UserModel
	if err := tx.SelectContext(ctx, &userModels, "SELECT id, name FROM users"); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	users := make([]User, len(userModels))
	for i := range userModels {
		user, err := fillUserResponse(ctx, tx, *userModels[i])
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
		users[i] = user
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, users)
}

func verifyUserSession(c echo.Context) error {
	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	sessionExpires, ok := sess.Values[defaultSessionExpiresKey]
	if !ok {
		// FIXME: エラーメッセージを検討する
		return echo.NewHTTPError(http.StatusForbidden, "")
	}

	now := time.Now()
	if now.Unix() > sessionExpires.(int64) {
		return echo.NewHTTPError(http.StatusUnauthorized, "session has expired")
	}

	return nil
}

func fillUserResponse(ctx context.Context, tx *sqlx.Tx, userModel UserModel) (User, error) {
	themeModel := ThemeModel{}
	if err := tx.GetContext(ctx, &themeModel, "SELECT * FROM themes WHERE user_id = ?", userModel.ID); err != nil {
		return User{}, err
	}

	user := User{
		ID:          userModel.ID,
		Name:        userModel.Name,
		DisplayName: userModel.DisplayName,
		Description: userModel.Description,
		CreatedAt:   userModel.CreatedAt,
		UpdatedAt:   userModel.UpdatedAt,
		Theme: Theme{
			ID:        themeModel.ID,
			DarkMode:  themeModel.DarkMode,
			CreatedAt: themeModel.CreatedAt,
		},
	}

	return user, nil
}

File: webapp/node/src/handlers/user-handler.ts
import { Context } from 'hono'
import { ResultSetHeader, RowDataPacket } from 'mysql2/promise'
import { HonoEnvironment } from '../types/application'
import {
  defaultUserIDKey,
  defaultUserNameKey,
  defaultSessionExpiresKey,
} from '../contants'
import { verifyUserSessionMiddleware } from '../middlewares/verify-user-session-middleare'
import { fillUserResponse } from '../utils/fill-user-response'
import { throwErrorWith } from '../utils/throw-error-with'
import { IconModel, UserModel } from '../types/models'

// GET /api/user/:username/icon
export const getIconHandler = [
  async (c: Context<HonoEnvironment, '/api/user/:username/icon'>) => {
    const username = c.req.param('username')

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[user]] = await conn
        .query<(UserModel & RowDataPacket)[]>(
          'SELECT * FROM users WHERE name = ?',
          [username],
        )
        .catch(throwErrorWith('failed to get user'))

      if (!user) {
        await conn.rollback()
        return c.text('not found user that has the given username', 404)
      }

      const [[icon]] = await conn
        .query<(Pick<IconModel, 'image'> & RowDataPacket)[]>(
          'SELECT image FROM icons WHERE user_id = ?',
          [user.id],
        )
        .catch(throwErrorWith('failed to get icon'))
      if (!icon) {
        await conn.rollback()
        return c.body(await c.get('runtime').fallbackUserIcon(), 200, {
          'Content-Type': 'image/jpeg',
        })
      }

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.body(icon.image, 200, {
        'Content-Type': 'image/jpeg',
      })
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// POST /api/icon
export const postIconHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/icon'>) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware

    // base64 encoded image
    const body = await c.req.json<{ image: string }>()

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      await conn
        .execute('DELETE FROM icons WHERE user_id = ?', [userId])
        .catch(throwErrorWith('failed to delete old user icon'))

      const [{ insertId: iconId }] = await conn
        .query<ResultSetHeader>(
          'INSERT INTO icons (user_id, image) VALUES (?, ?)',
          [userId, Buffer.from(body.image, 'base64')],
        )
        .catch(throwErrorWith('failed to insert icon'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json({ id: iconId }, 201)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// GET /api/user/me
export const getMeHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/user/me'>) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[user]] = await conn
        .query<(UserModel & RowDataPacket)[]>(
          'SELECT * FROM users WHERE id = ?',
          [userId],
        )
        .catch(throwErrorWith('failed to get user'))

      if (!user) {
        await conn.rollback()
        return c.text('not found user that has the userid in session', 404)
      }

      const response = await fillUserResponse(
        conn,
        user,
        c.get('runtime').fallbackUserIcon,
      ).catch(throwErrorWith('failed to fill user'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(response)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// ユーザ登録API
// POST /api/register
export const registerHandler = async (
  c: Context<HonoEnvironment, '/api/register'>,
) => {
  const body = await c.req.json<{
    name: string
    display_name: string
    description: string
    password: string
    theme: { dark_mode: boolean }
  }>()

  if (body.name === 'pipe') {
    return c.text("the username 'pipe' is reserved", 400)
  }

  const hashedPassword = await c
    .get('runtime')
    .hashPassword(body.password)
    .catch(throwErrorWith('failed to generate hashed password'))

  const conn = await c.get('pool').getConnection()
  await conn.beginTransaction()

  try {
    const [{ insertId: userId }] = await conn
      .execute<ResultSetHeader>(
        'INSERT INTO users (name, display_name, description, password) VALUES(?, ?, ?, ?)',
        [body.name, body.display_name, body.description, hashedPassword],
      )
      .catch(throwErrorWith('failed to insert user'))

    await conn
      .execute('INSERT INTO themes (user_id, dark_mode) VALUES(?, ?)', [
        userId,
        body.theme.dark_mode,
      ])
      .catch(throwErrorWith('failed to insert user theme'))

    await c
      .get('runtime')
      .exec([
        'pdnsutil',
        'add-record',
        'u.isucon.dev',
        body.name,
        'A',
        '0',
        c.get('runtime').powerDNSSubdomainAddress,
      ])
      .catch(throwErrorWith('failed to add record to powerdns'))

    const response = await fillUserResponse(
      conn,
      {
        id: userId,
        name: body.name,
        display_name: body.display_name,
        description: body.description,
      },
      c.get('runtime').fallbackUserIcon,
    ).catch(throwErrorWith('failed to fill user'))

    await conn.commit().catch(throwErrorWith('failed to commit'))

    return c.json(response, 201)
  } catch (error) {
    await conn.rollback()
    return c.text(`Internal Server Error\n${error}`, 500)
  } finally {
    await conn.rollback()
    conn.release()
  }
}

// ユーザログインAPI
// POST /api/login
export const loginHandler = async (
  c: Context<HonoEnvironment, '/api/login'>,
) => {
  const body = await c.req.json<{
    username: string
    password: string
  }>()

  const conn = await c.get('pool').getConnection()
  await conn.beginTransaction()

  try {
    // usernameはUNIQUEなので、whereで一意に特定できる
    const [[user]] = await conn
      .query<(UserModel & RowDataPacket)[]>(
        'SELECT * FROM users WHERE name = ?',
        [body.username],
      )
      .catch(throwErrorWith('failed to get user'))

    if (!user) {
      await conn.rollback()
      return c.text('invalid username or password', 401)
    }

    await conn.commit().catch(throwErrorWith('failed to commit'))

    const isPasswordMatch = await c
      .get('runtime')
      .comparePassword(body.password, user.password)
      .catch(throwErrorWith('failed to compare hash and password'))
    if (!isPasswordMatch) {
      return c.text('invalid username or password', 401)
    }

    // 1時間でセッションが切れるようにする
    const sessionEndAt = Date.now() + 1000 * 60 * 60

    const session = c.get('session')
    session.set(defaultUserIDKey, user.id)
    session.set(defaultUserNameKey, user.name)
    session.set(defaultSessionExpiresKey, sessionEndAt)

    // eslint-disable-next-line unicorn/no-null
    return c.body(null)
  } catch (error) {
    await conn.rollback()
    return c.text(`Internal Server Error\n${error}`, 500)
  } finally {
    await conn.rollback()
    conn.release()
  }
}

// GET /api/user/:username
export const getUserHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/user/:username'>) => {
    const username = c.req.param('username')

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[user]] = await conn
        .query<(UserModel & RowDataPacket)[]>(
          'SELECT * FROM users WHERE name = ?',
          [username],
        )
        .catch(throwErrorWith('failed to get user'))

      if (!user) {
        await conn.rollback()
        return c.text('not found user that has the given username', 404)
      }

      const response = await fillUserResponse(
        conn,
        user,
        c.get('runtime').fallbackUserIcon,
      ).catch(throwErrorWith('failed to fill user'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(response)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]
